### PR TITLE
[4.0] Plugin for managing Atum settings

### PR DESCRIPTION
**Pull request by SharkyKZ**
### Summary of Changes

Currently Atum params are hardcoded in com_users form. Unless we expect every other template to follow the same settings, this is totally wrong.

This moves those settings to a separate plugin.

### Testing Instructions

Edit your profile in backend, change some Atum settings.
Check that they're still respected.

### Expected result

Works like before.

### Documentation Changes Required

IDK

### TODO
- [x] SQL files

### CODE REVIEW

**Comment by brianteeman** [comment]
>  Unless we expect every other template to follow the same settings, this is totally wrong.

Yes I would like the two or three other possible admin templates to follow the same settings. My 2c

File: administrator/templates/atum/index.php [hunk 1]
```diff
@@ -33,10 +33,13 @@
...
-$a11y_mono      = (bool) $app->getIdentity()->getParam('a11y_mono', '');
-$a11y_contrast  = (bool) $app->getIdentity()->getParam('a11y_contrast', '');
-$a11y_highlight = (bool) $app->getIdentity()->getParam('a11y_highlight', '');
-$a11y_font      = (bool) $app->getIdentity()->getParam('a11y_font', '');
+$user       = $app->getIdentity();
+$contrast   = (bool) $user->getParam('atum.contrast', 0);
+$highlight  = (bool) $user->getParam('atum.highlight', 0);
+$fontsize   = (bool) $user->getParam('atum.fontsize', 0);
+
+// Monochrome can also bet set in the template.
```
Comment (Quy) [inline administrator/templates/atum/index.php]:
```suggestion
// Monochrome can also be set in the template.
```

File: administrator/templates/atum/templateDetails.xml [hunk 2]
```diff
@@ -110,6 +110,8 @@
 					label="TPL_ATUM_COLORS_SETTINGS_MONOCHROME_LABEL"
 					layout="joomla.form.field.radio.switcher"
 					default="0"
+					filter="integer"
+					validate="options"
```
Comment (Quy) [inline administrator/templates/atum/templateDetails.xml:173]:
```suggestion
```
Options validation is for list type.

Comment (SharkyKZ) [reply administrator/templates/atum/templateDetails.xml:173]: Radio is a list-based field.

Comment (Quy) [reply administrator/templates/atum/templateDetails.xml:173]: Per documentation:

> Requires the value entered be one of the options in an element of type="list": that is, that the element is a select list.

Comment (SharkyKZ) [reply administrator/templates/atum/templateDetails.xml:173]: It can be any field extending from `Joomla\CMS\Form\Field\ListField`.

**Comment by mbabker** [comment]
> > Unless we expect every other template to follow the same settings, this is totally wrong.
> 
> Yes I would like the two or three other possible admin templates to follow the same settings. My 2c

Then they need to be advertised as system settings.  If they are template related settings for a specific template, then core should be following the same practices that the ecosystem must follow, and the ecosystem doesn't get to stuff their settings directly into a core component.

**Comment by dgrammatiko** [comment]
> Then they need to be advertised as system settings

That! Also rename it to `Accessibility options` and name them carefully

PS. @C-Lodder already mentioned it elsewhere that these settings **ARE OS** wide and thus there is no need to pamper the users with such plugins (the js thingy that renders the setting this plugin is moving around). Also it's a very good idea to ask some people with disabilities if such things are helpful to them (the answer is **no** because they use the browsers css override or other higher level solutions. In sort Joomla is solving a non existing problem, not the first instance...)

**Comment by Quy** [comment]
Remove `Accessibility Settings` from `User Menu` dropdown when plugin is disabled.

File: administrator/modules/mod_user/tmpl/default.php [hunk 3]
```diff
@@ -42,11 +42,6 @@
 			<span class="fas fa-fw fa-user" aria-hidden="true"></span>
 			<?php echo Text::_('MOD_USER_EDIT_ACCOUNT'); ?>
 		</a>
-		<?php $route = 'index.php?option=com_users&task=user.edit&id=' . $user->id . '&return=' . base64_encode($uri) . '#attrib-accessibility'; ?>
-		<a class="dropdown-item" href="<?php echo Route::_($route); ?>">
-			<span class="fas fa-fw fa-universal-access" aria-hidden="true"></span>
-			<?php echo Text::_('MOD_USER_ACCESSIBILITY_SETTINGS'); ?>
```
Comment (Quy) [inline administrator/modules/mod_user/tmpl/default.php:48]: Remove language string.

**Comment by HLeithner** [comment]
I see no reason to move a11y settings for the template out of the user options into a plugin, basically every template should use them (at least in Europe).

**Comment by brianteeman** [comment]
@HLeithner actually they could use the plugin instead

**Comment by dgrammatiko** [comment]
In my previous comment I've mentioned that disabled people are using higher level solutions (eg Browser plugins). Here is the proof, dyslexic people override the fonts: https://speakerdeck.com/ninjanails/death-to-icon-fonts 

Don't try to solve non existing problems because someone **assumes**. COVID19 teach us something: read and the data and act upon them, Joomla didn't even collected any data to make such wild assumptions...


PS if you want to make Joomla truly **ACCESSIBLE** then **DROP** the Font Awesome ICON and use their svgs. You're already **inaccessible** for dyslexic people that will override the fonts.
In sort solve real problems not unicorn ones...

**Comment by brianteeman** [comment]
Sorry but that is not 100% correct - I was under the same misapprehension myself until relatively recently. People with **permanent** disabilities will of course try to implement higher level solutions but not all disabilities are permanent.

**Comment by dgrammatiko** [comment]
@brianteeman all I'm saying is Joomla has bigger fish to fry to be accessible (eg the icon font!!). The widget is questionable. AFAIK gov.uk is totally accessible **without** any fancy widget

**Comment by C-Lodder** [comment]
The below refers to the colour settings only:

Permanent or not, if you have a disability, you don't go to the settings for each individual site you use and change to greyscale, high contrast, etc.
You use software or change at OS level (normally the latter)
Even some modern monitors have these types of settings.

Windows 10, MacOS, Android and iOS all have colour filters:

- Dark mode
- Greyscale
- High contrast
- Protanopia
- Deuteranopia
- Tritanopia

By all means, keep whem if you want, but if you're going to do it, do it properly and support the 4 other critical ones.

**Comment by brianteeman** [comment]
I am not commenting further on this as you have misunderstood the objective of these settings.

**Comment by C-Lodder** [comment]
Enlighten me on the objective of them then...

File: plugins/user/atum/atum.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function onContentPrepareForm(Form $form, $data)
+	{
+		// Check we are manipulating the correct form.
+		if (!in_array($form->getName(), ['com_admin.profile', 'com_users.user'], true))
```
Comment (bembelimen) [inline plugins/user/atum/atum.php]: Does it make sense to check for some admin stuff? So it's only visible in the frontend when I have admin access + the template is active?

Comment (SharkyKZ) [reply plugins/user/atum/atum.php]: Yes, it does.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 833744e936a6da7ac2d5cc643fb811eacc0fa3c4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29226">issues.joomla.org/tracker/joomla-cms/29226</a>.</sub>

**Comment by brianteeman** [comment]
Please add to the list of core extensions

File: plugins/user/atum/atum.php (hunk 4 above)
Comment (bembelimen) [reply plugins/user/atum/atum.php]: Thx

**Comment by Quy** [comment]
Should these settings be displayed/editable on the frontend in edit profile since they are only applicable on the backend?

**Comment by SharkyKZ** [comment]
Maybe. Users can select backend template and language in frontend.

**Comment by Quy** [comment]
Please fix conflicts.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on df8581895ba73352bf7c8f823eb0e454c8611b9e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29226">issues.joomla.org/tracker/joomla-cms/29226</a>.</sub>

**Comment by Quy** [comment]
Please fix display value of `1`.

**Comment by Fedik** [comment]
Just an idea:
Alternatively to use the plugin, we could extend `templateDetails.xml` markup to have something like `<userParams>` tag with fieldsets of User specific settings, and then User model will load it to an user Params form. This will work across all templates without extra plugin.

File: administrator/language/en-GB/plg_user_atum.ini [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+; License GNU General Public License version 2 or later; see LICENSE.txt
+; Note : All ini files need to be saved as UTF-8
+
+PLG_USER_ATUM="User - Atum"
+PLG_USER_ATUM_SETTINGS_FIELD_CONTRAST="High Contrast"
+PLG_USER_ATUM_SETTINGS_FIELD_FONTSIZE="Increase Font Size"
+PLG_USER_ATUM_SETTINGS_FIELD_HIGHLIGHT="Highlight Links"
+PLG_USER_ATUM_SETTINGS_FIELD_MONOCHROME="Monochrome"
+PLG_USER_ATUM_SETTINGS_FIELDSET_LABEL="Atum Template Settings"
+PLG_USER_ATUM_DEFAULT_VALUE="Website default"
```
Comment (Quy) [inline administrator/language/en-GB/plg_user_atum.ini:8]: Alpha order.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 965a652cda7c650251923532a80de38749c904e5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29226">issues.joomla.org/tracker/joomla-cms/29226</a>.</sub>

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on 965a652cda7c650251923532a80de38749c904e5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29226">issues.joomla.org/tracker/joomla-cms/29226</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29226">issues.joomla.org/tracker/joomla-cms/29226</a>.</sub>

**Comment by brianteeman** [comment]
Want to go on record saying that for me the renaming of this plugin (removing any connection to accessibility) is wrong.
Removing the resulting accessibility option link from the user profile is wrong (that was in the very first design of j4)

As it is right now it makes no sense. The description is that this is for per user settings of the atum template so I would therefore expect to see the same options in the atum template itself for default settings and vice versa.

This is just wrong.

**Comment by SharkyKZ** [comment]
>Want to go on record saying that for me the renaming of this plugin (removing any connection to accessibility) is wrong.

Language strings can be changed.

>Removing the resulting accessibility option link from the user profile is wrong (that was in the very first design of j4)

This can be added in a layout override for Atum.

>The description is that this is for per user settings of the atum template so I would therefore expect to see the same options in the atum template itself

These can be added. Monochrome option is already present in template params.

**Comment by SharkyKZ** [comment]
@Fedik make a PR if you want.

**Comment by brianteeman** [comment]
If you have to say "these can be added" then this can not be RTC. Don't just remove functionality and expect someone else to put it back.

**Comment by Fedik** [comment]
I think we need to have a common decision on this topic before doing changes here, instead of when everyone doing something own.

I found that my idea also not much good, because then we need to store the params for every template style,
 eg `$userParams[template][blablaTemplate][styleId] = [.. user params for the template style...]`

For now, I think, I would keep it as it is now (in User params), there no big reason to rush it while the beta state. We can back to the topic in a future, in one of `4.x`

**Comment by SharkyKZ** [comment]
PR updated.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29226">issues.joomla.org/tracker/joomla-cms/29226</a>.</sub>

File: administrator/templates/atum/index.php [hunk 6]
```diff
@@ -29,11 +29,12 @@
 $cpanel     = $option === 'com_cpanel';
 $hiddenMenu = $app->input->get('hidemainmenu');
 
-// Getting user accessibility settings
-$a11y_mono      = (bool) $app->getIdentity()->getParam('a11y_mono', '');
-$a11y_contrast  = (bool) $app->getIdentity()->getParam('a11y_contrast', '');
-$a11y_highlight = (bool) $app->getIdentity()->getParam('a11y_highlight', '');
-$a11y_font      = (bool) $app->getIdentity()->getParam('a11y_font', '');
+// Get accessibility settings
+$user       = $app->getIdentity();
```
Comment (brianteeman) [inline administrator/templates/atum/index.php:33]: This is not an accessibility setting so the comment above it makes no sense

**Comment by brianteeman** [comment]
https://github.com/joomla/joomla-cms/pull/29226#issuecomment-657243245

File: administrator/templates/atum/index.php (hunk 6 above)
Comment (SharkyKZ) [reply administrator/templates/atum/index.php:33]: The comment is for the code block, not just for this specific line.

**Comment by HLeithner** [comment]
since we #31884 is merged we don't need this pr any longer thanks for your time creating this anyway!

**Comment by SharkyKZ** [comment]
#31884 is not related to this.

**Pull request closed without merging**